pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,263 @@
+	rw := revisionWatcher{
+		rev:          rev,
+		updateCb:     updateCb,
+		healthyDests: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+	return &rw
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		// TODO(greghaynes) log error here
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, fmt.Sprintf("%s:%d", addr.IP, portVal))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := []string{}
+	for dest, healthy := range dests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+func (rw *revisionWatcher) checkDests() {
+	healthyDests := make(map[string]bool)
+	var healthyDestsMux sync.Mutex
+	var probeWg sync.WaitGroup
+
+	// Context used for our requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthyDests[dest]; ok && curHealthy {
+			healthyDests[dest] = true
+		} else {
+			probeWg.Add(1)
+			go func(dest string) {
+				httpDest := url.URL{
+					Scheme: "http",
+					Host:   dest,
+				}
+				ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+					prober.WithHeader(network.ProbeHeaderName, queue.Name),
+					prober.ExpectsBody(queue.Name))
+
+				healthyDestsMux.Lock()
+				healthyDests[dest] = ok
+				healthyDestsMux.Unlock()
+
+				if err != nil {
+					rw.logger.Errorw(fmt.Sprintf("Probing of dest %q failed", dest), zap.Error(err))
+				} else if !ok {
+					rw.logger.Infof("Probing of dest %q failed.", dest)
+				}
+
+				probeWg.Done()
+			}(dest)
+		}
+	}
+
+	probeWg.Wait()
+
+	rw.healthyDests = healthyDests
+	destsArr := filterHealthyDests(healthyDests)
+	rw.updateCb(rw.rev, destsArr)
+}
+
+func (rw *revisionWatcher) runOnce(tickCh <-chan time.Time) bool {
+	select {
+	case dests := <-rw.destsChan:
+		if dests == nil {
+			// shutdown
+			return true
+		}
+		rw.dests = dests
+		rw.checkDests()
+	case <-tickCh:
+		rw.checkDests()
+	}
+	return false
+}
+
+func (rw *revisionWatcher) run() {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if rw.runOnce(ticker.C) {
+			return
+		}
+	}
+}
+
+// RevisionBackendsManager listens to revision endpoints and keeps track of healthy
+// l4 dests which can be used to reach a revisoin
+type RevisionBackendsManager struct {
+	revisionWatchers    map[RevisionID]*revisionWatcher
+	revisionDestsCh     map[RevisionID]chan []string
+	revisionWatchersMux sync.RWMutex
```

Comment:
This guards both the revisionWatchers and the revisionDestsCh, right? Should we fold them into one map and store a struct that hold both the `revisionWatcher` and the channel? Less bookkeeping I'd think.